sql-plugin/src/main/311+-all/scala/com/nvidia/spark/ParquetCachedBatchSerializer.scala

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark
+
+import com.nvidia.spark.rapids.ShimLoader
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, Expression}
+import org.apache.spark.sql.columnar.{CachedBatch, CachedBatchSerializer}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.storage.StorageLevel
+
+
+trait GpuCachedBatchSerializer extends CachedBatchSerializer {
+  def gpuConvertCachedBatchToColumnarBatch(
+      input: RDD[CachedBatch],
+      cacheAttributes: Seq[Attribute],
+      selectedAttributes: Seq[Attribute],
+      conf: SQLConf): RDD[ColumnarBatch]
+}
+
+/**
+ * User facing wrapper class that calls into the proper shim version.
+ */
+class ParquetCachedBatchSerializer extends GpuCachedBatchSerializer {
+
+  val minSupportedVer = "3.1.1"
+  val sparkVersion = ShimLoader.getSparkVersion
+  // Note that since the config to set the serializer wasn't added until
+  // Spark 3.1.0 (https://issues.apache.org/jira/browse/SPARK-32274) this shouldn't
+  // ever throw.
+  if (sparkVersion < minSupportedVer) {
+    throw new IllegalArgumentException("ParquetCachedBaatchSerializer only supported for Spark " +
+      s"versions > 3.1.1, version found was: $sparkVersion")
+  }
+
+  private lazy val realSerializer: GpuCachedBatchSerializer = {
+    ShimLoader.newInstanceOf("com.nvidia.spark.rapids.shims.v2.ParquetCachedBatchSerializer")
```

Comment:
Can we please move the reflection to the Shim loader? For me it makes it simpler to have all of the reflection in a single place so we know exactly how much is being used and what it is used for.

Reply:
I mentioned this in the description, we can't because the class only existing in spark 3.1.1 and we have to pull a common version of Shimloader which is 3.0.1

Reply:
By class I mean CachedBatchSErializer which comes from spark

Reply:
or we could create a Shimloader version for 311+

Reply:
That stinks, but I understand why.